### PR TITLE
Progress for adding files

### PR DESCRIPTION
**Pull request by rklaehn**
Use an indicativ multiprogress bar.

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -44,6 +44,7 @@ ssh-key = { version = "0.5.1", features = ["ed25519", "std", "rand_core"] }
 tempfile = "3.4"
 thiserror = "1"
 tokio = { version = "1", features = ["full"] }
+tokio-stream = "0.1.12"
```
Comment (flub) [inline Cargo.toml:47]: is `0.1` enough?

File: src/main.rs [hunk 2]
```diff
@@ -201,6 +202,125 @@ async fn make_rpc_client(
...
+    pub size: u64,
+}
+
+async fn aggregate_add_response(
+    stream: impl Stream<
+            Item = std::result::Result<
+                ProvideProgress,
+                impl std::error::Error + Send + Sync + 'static,
+            >,
+        > + Unpin,
```
Comment (flub) [inline src/main.rs:262]: nested impl traits!  could consider moving this to a where clause

File: src/main.rs [hunk 3]
```diff
@@ -201,6 +202,125 @@ async fn make_rpc_client(
...
+) -> anyhow::Result<(Hash, Vec<ProvideResponseEntry>)> {
+    let mut stream = stream;
```
Comment (flub) [inline src/main.rs:264]: you can't declare this mut in the signature?

File: src/util.rs [hunk 4]
```diff
@@ -242,3 +244,67 @@ mod tests {
         assert_eq!(canonicalize_path("foo/bar").unwrap(), "foo/bar");
     }
 }
+
+pub(crate) struct ProgressReader<R, F: Fn(ProgressReaderUpdate)> {
```
Comment (flub) [inline src/util.rs:248]: i assume you looked at `crate::progress::ProgressEmitter` and didn't find it suitable?

**Review by flub — APPROVED** [review]

File: src/main.rs (hunk 3 above)
Comment (rklaehn) [reply src/main.rs:264]: I can, but I don't want to. I find that the mut is entirely an implementation detail and does not belong in the signature.

Comment (flub) [reply src/main.rs:264]: putting `mut` in the "signature" has no effect on the caller at all.  It's a weirdness of Rust syntax where putting `mut` there or not has no effect on the signature at all, it is just something local to the function body.

So you're saying you so strongly disagree with Rust's syntax choices that you prefer to do manual work arounds :smile:  While I agree it's a bit odd, I don't tend to agree with fighting stuff like that.  Once you know how this works I think it's perfectly fine.

**Pull request merged**